packages/react/src/components/gdpr-consent-dialog/state.ts

```diff
@@ -14,10 +16,21 @@ type Action =
   | { type: "STOP_LOADING" }
   | { type: "SET_HAS_ERROR"; payload: boolean }
   | { type: "SET_IS_CUSTOMIZING"; payload: boolean }
+  | { type: "SET_CONSENT_SETTINGS"; payload: GDPRConsent[] }
   | { type: "TOGGLE_CONSENT"; payload: ConsentSettingsLevel };
 
 export type Dispatch = React.Dispatch<Action>;
 
+export const mapConsentsToSettings = (consents: GDPRConsent[]) => {
+  const settings = Object.fromEntries(
+    CONSENT_LEVELS_WITHOUT_NONE.map((level) => [
+      level,
+      consents.map(({ consent_level }) => consent_level).includes(level),
+    ])
+  );
+  return settings as ConsentSettings;
```

Comment:
How come the cast is necessary here? Casting like this is usually a code smell

Reply:
`Object.fromEntries` here doesn't preserve the key's type and simply returns a key as string so a cast is necessary, do you know a way to handle this without casting?

Reply:
Ah I see, okay

There's two things we would need to do to remove the cast:
- Type the map correctly so it's clear the return type is a tuple
- Create a typed wrapper for `Object.fromEntries`

// util.ts
export const fromEntries = <const T extends ReadonlyArray<readonly [PropertyKey, unknown]>>(
  entries: T
): { [K in T[number] as K[0]]: K[1] } => {
  return Object.fromEntries(entries) as { [K in T[number] as K[0]]: K[1] };
};
```

```
// state.ts
import { fromEntries } from './util.ts'

const settings: ConsentSettings = fromEntries(
  CONSENT_LEVELS_WITHOUT_NONE.map<[typeof CONSENT_LEVELS_WITHOUT_NONE[number], boolean]>((level) => [
    level,
    consents.map(({ consent_level }) => consent_level).includes(level),
  ])
)
```

Now we have type safety